##################################################
## Project: BRB-seq
## Script purpose: Generating Figure 3, Panel D
## Date: 2022 Oct 04
## Author: Vincent Gardeux (vincent.gardeux@epfl.ch)
##################################################

## Change working directory
#setwd("")

## Load libraries
suppressPackageStartupMessages(require(plyr))
suppressPackageStartupMessages(require(ggplot2))

## Functions
# Generate Downsampled names
downsample.names <- function(name, cnt, nreps){
  res = c()
  for(n in 1:nreps){
    res = c(res, paste0(name,".DS_",cnt,"_",n))
  }
  res
}

# Generate sd and mean for computing error bars
data_summary <- function(data, varname, groupnames){
  summary_func <- function(x, col){
    c(mean = mean(x[[col]], na.rm=TRUE), sd = sd(x[[col]], na.rm=TRUE))
  }
  data_sum<-ddply(data, groupnames, .fun=summary_func, varname)
  data_sum <- rename(data_sum, c("mean" = varname))
  return(data_sum)
}

## Load Truseq Downsampled data (downsampling was performed 5 times on aligned reads from BAM files)
data.raw = read.table("truseq.count.matrix.SE.DS.txt", sep="\t", header = T, row.names=1, check.names = F)
# Remove extra rows generated by HTseq-count
remove = c("__no_feature", "__ambiguous", "__too_low_aQual", "__not_aligned", "__alignment_not_unique")
data.raw = data.raw[!(rownames(data.raw) %in% remove),]
# Restrict to samples with 1M reads (downsampling was performed 5 times on aligned reads from BAM files, i.e. 1M includes the 'remove' part that we removed just before)
data.raw = data.raw[,c(downsample.names("T0A",1000,5), downsample.names("T0B",1000,5), downsample.names("T14A",1000,5), downsample.names("T14B",1000,5))]
# Compute number of detected genes with different thresholds, and store in global data.frame
data.barplot <- data.frame(reads=c(colSums(data.raw > 0), colSums(data.raw > 1), colSums(data.raw > 100)), level=c(rep(">0", 20), rep(">1", 20), rep(">100", 20)), type="TruSeq")

## Load BRB-seq Downsampled data (downsampling was performed 5 times on aligned reads from BAM files)
data.raw = read.table("brbseq.count.matrix.DS.txt", sep="\t", header = T, row.names=1, check.names = F)
# Remove extra rows generated by BRBseq-Tools
remove = c("__no_feature", "__ambiguous", "__too_low_aQual", "__not_aligned", "__alignment_not_unique")
data.raw = data.raw[!(rownames(data.raw) %in% remove),]
# Restrict to samples with 1M reads (downsampling was performed 5 times on aligned reads from BAM files, i.e. 1M includes the 'remove' part that we removed just before)
data.raw = data.raw[,c(downsample.names("DA4_t0A",1000,5), downsample.names("DA4_t0B",1000,5), downsample.names("DA4_t14A",1000,5), downsample.names("DA4_t14B",1000,5))]
# Compute number of detected genes with different thresholds, and add to previous global data.frame
data.barplot <- rbind(data.barplot, data.frame(reads=c(colSums(data.raw > 0), colSums(data.raw > 1), colSums(data.raw > 100)), level=c(rep(">0", 20), rep(">1", 20), rep(">100", 20)), type="BRB-seq"))

## Compute SD & generate error bars
df3 <- data_summary(data.barplot, varname="reads", groupnames=c("level", "type"))
df3$sd_min <- df3$reads - df3$sd
df3$sd_max <- df3$reads + df3$sd
# It's a stacked barplot, so I need to remvove the extra layers, so the sum is everything >0
df3[df3$level == ">1", "reads"] <- df3[df3$level == ">1", "reads"] - df3[df3$level == ">100", "reads"]
df3[df3$level == ">0", "reads"] <- df3[df3$level == ">0", "reads"] - df3[df3$level == ">100", "reads"] - df3[df3$level == ">1", "reads"]

# Convert dose to a factor variable
df3$level=factor(df3$level, levels = c(">0", ">1", ">100"))

# Plot
p <- ggplot(df3, aes(x = type, y = reads, fill = level)) 
p <- p + geom_bar(stat="identity", position="stack") 
p <- p + geom_errorbar(aes(ymin=sd_min, ymax=sd_max), width=.1, size = .2)
p <- p + ylab("Number of detected genes") + theme_bw() + theme(legend.position = "right", legend.justification = "bottom", axis.ticks.x = element_blank(), axis.line.y = element_line(), panel.border = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.title.x = element_blank(), axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, face="bold")) 
p <- p + scale_y_continuous(breaks = seq(0, 15000, 5000), limits = c(0,15000), expand = c(0,0))  + scale_fill_discrete(name = "Reads", type = c("#D4D4D4", "#909090", "#494848"))
ggsave(p, filename = "Figure3D.pdf", width = 3, height = 6)
