##################################################
## Project: BRB-seq
## Script purpose: Generating Figure 3, Panel C
## Date: 2022 Oct 04
## Author: Vincent Gardeux (vincent.gardeux@epfl.ch)
##################################################

## Change working directory
#setwd("")

## Load libraries
suppressPackageStartupMessages(require(reshape2)) # for using the 'melt' function
suppressPackageStartupMessages(require(ggplot2))

## Functions
percent <- function(x, digits = 2, format = "f", ...) {
  paste0(formatC(x, format = format, digits = digits, ...), "%")
}

## Load Truseq data
data.alignment = read.table("data/truseq.alignment.summary.SE.txt", sep="\t", header = T, row.names=1, check.names = F)
data.count = read.table("data/truseq.count.matrix.SE.txt", sep="\t", header = T, row.names=1, check.names = F)
# Remove extra rows generated by HTseq-count
remove = c("__no_feature", "__ambiguous", "__too_low_aQual", "__not_aligned", "__alignment_not_unique")
data.count = data.count[!(rownames(data.count) %in% remove),]
# Create data.frame to use for final plot
data.barplot = data.frame(reads.demult = sum(data.alignment$input.reads), uniquely.mapped = sum(data.alignment$uniquely.mapped), mapped.to.genes = sum(colSums(data.count)))
data.barplot$uniquely.mapped = data.barplot$uniquely.mapped - data.barplot$mapped.to.genes
data.barplot$reads.demult = data.barplot$reads.demult - data.barplot$uniquely.mapped - data.barplot$mapped.to.genes
data.barplot$type = "TruSeq"

## Load BRB-seq data
data.alignment = read.table("data/brbseq.alignment.summary.txt", sep="\t", header=TRUE, row.names=1, check.names = F)
data.count = read.table("data/brbseq.count.matrix.txt", sep="\t", header = T, row.names=1, check.names = F)
# Remove extra rows generated by HTseq-count
remove = c("__no_feature", "__ambiguous", "__too_low_aQual", "__not_aligned", "__alignment_not_unique")
data.count = data.count[!(rownames(data.count) %in% remove),]
# Create data.frame to use for final plot
data.barplot.tmp = data.frame(reads.demult = sum(data.alignment.da4$input.reads), uniquely.mapped = sum(data.alignment.da4$uniquely.mapped), mapped.to.genes = sum(colSums(data.count.da4)))
data.barplot.tmp$uniquely.mapped = data.barplot.tmp$uniquely.mapped - data.barplot.tmp$mapped.to.genes
data.barplot.tmp$reads.demult = data.barplot.tmp$reads.demult - data.barplot.tmp$uniquely.mapped - data.barplot.tmp$mapped.to.genes
data.barplot.tmp$type = "BRB-seq"
data.barplot = rbind(data.barplot, data.barplot.tmp)

## Plot
# In percent
total = data.barplot$mapped.to.genes + data.barplot$uniquely.mapped + data.barplot$reads.demult
data.barplot$mapped.to.genes = (data.barplot$mapped.to.genes / total) * 100
data.barplot$uniquely.mapped = (data.barplot$uniquely.mapped / total) * 100
data.barplot$reads.demult = (data.barplot$reads.demult/ total) * 100
data.barplot <- melt(data.barplot, id = "type")
data.barplot$label <- round(data.barplot$value,1)
# Make labels
data.barplot[data.barplot$variable == "uniquely.mapped" & data.barplot$type == "TruSeq", "label"] <- data.barplot[data.barplot$variable == "uniquely.mapped" & data.barplot$type == "TruSeq", "label"] + data.barplot[data.barplot$variable == "mapped.to.genes" & data.barplot$type == "TruSeq", "label"]
data.barplot[data.barplot$variable == "uniquely.mapped" & data.barplot$type == "BRB-seq", "label"] <- data.barplot[data.barplot$variable == "uniquely.mapped" & data.barplot$type == "BRB-seq", "label"] + data.barplot[data.barplot$variable == "mapped.to.genes" & data.barplot$type == "BRB-seq", "label"]
data.barplot[data.barplot$variable == "reads.demult" & data.barplot$type == "TruSeq", "label"] <- ""
data.barplot[data.barplot$variable == "reads.demult" & data.barplot$type == "BRB-seq", "label"] <- ""
# Plot
p <- ggplot(data.barplot, aes(x = type, y = value, fill = variable)) + geom_bar(stat = "identity", position="stack") + ylab("Number of reads (%)") + theme_bw() + theme(legend.position = "right", legend.justification = "bottom", axis.ticks.x = element_blank(), axis.line.y = element_line(), panel.border = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.title.x = element_blank(), axis.text.x = element_text(angle = 45, vjust = 1, hjust=1, face="bold")) + scale_y_continuous(breaks = seq(0, 100, 20), limits = c(0,100), expand = c(0,0))  + scale_fill_discrete(name = "Alignment", labels=c("No / multiple", "No gene / ambiguous", "Mapped to genes"), type = c("#D4D4D4", "#909090", "#494848")) + geom_text(aes(label = label), size = 4, color = "white", fontface = "bold", position = position_stack(vjust = 0.8))
ggsave(p, filename = "Figure3C.pdf", width = 4, height = 6)
